docs/csharp/whats-new/csharp-8.md

```diff
@@ -0,0 +1,314 @@
+
+Nullable reference types aren't checked to ensure they aren't assigned or initialized to null. However, the compiler uses flow analysis to ensure that any variable of a nullable reference type is checked against null before it's accessed, or assigned to a nonnullable reference type.
+
+You can learn more about the feature in the overview of [nullable reference types](../nullable-references.md). Try it yourself in a new application in this [nullable reference types tutorial](../tutorials/nullable-reference-types.md). Learn about the steps to migrate an existing codebase to make use of nullable reference types in this tutorial for [migrating an application to use nullable reference types]../tutorials/upgrade-to-nullable-references.md).
+
+## Asynchronous streams
+
+Starting with C# 8.0, you can create and consume streams asynchronously. A method that returns an asynchronous stream has three properties:
+
+1. It was declared with the `async` modifier.
+1. It returns an `IAsyncEnumerable<T>`.
+1. The method contains `yield return` statements to return successive elements in the asynchronous stream.
+
+Consuming an asynchronous stream requires you to add the `await` keyword before the `foreach` keyword when you enumerate the elements of the stream. Adding the `await` keyword requires the method that enumerates the asynchronous stream was declared with the `async` modifier and returns a type allowed for an `async` method. Typically that means returning a <xref:System.Threading.Tasks.Task> or <xref:System.Threading.Tasks.Task%601>. It can also be a <xref:System.Threading.Tasks.ValueTask> or <xref:System.Threading.Tasks.ValueTask%601>. A method can both consume and produce an asynchronous stream, which means it would return an `IAsyncEnumerable<T>`.
+
+You can try asynchronous streams yourself in our tutorial on [creating and consuming async streams](../tutorials/generate-consume-asynchronous-stream.md).
+
+## Ranges and indices
+
+Ranges and indices provide a succinct syntax for specifying subranges in an array, <xref:System.Span%601>, or <xref:System.ReadonlySpan%601>. Consider the following array:
+
+```csharp
+var words = new string[]
+{
+    "The",
+    "quick",
+    "brown",
+    "fox",
+    "jumped",
+    "over",
+    "the",
+    "lazy",
+    "dog"
+};
+```
+
+You can retrieve the last word by prefixing the index `1` with `^`:
+
+```csharp
+Console.WriteLine($"The last word is {words[^1]}");
+```
+
+When you take only a single element and specify from the end, indexing starts at 1.
+
+You can also specify ranges of an array. The following code creates a subrange with the words "quick", "brown", and "fox":
+
+```csharp
+var brownFox = words[1..4];
```

Comment:
It seems confusing that this returns only 3 words, when you'd expect it to return the elements at indices 1 through 4, or 4 words An explanation of the syntax would help.